src/librustc_mir/interpret/eval_context.rs

```diff
@@ -767,11 +767,11 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         &self,
         gid: GlobalId<'tcx>,
     ) -> InterpResult<'tcx, MPlaceTy<'tcx, M::PointerTag>> {
-        // FIXME(oli-obk): make this check an assertion that it's not a static here
-        // FIXME(RalfJ, oli-obk): document that `Place::Static` can never be anything but a static
-        // and `ConstValue::Unevaluated` can never be a static
-        // FIXME(oli-obk, spastorino): the above FIXME is not true anymore, PlaceBase::Static does
-        // not exist anymore (except for promoteds but it's going away soon).
+        // For statics we pick `ParamEnv::reveal_all`, because statics don't have generics
+        // and thus don't care about the parameter environment. While we could just use
+        // `self.param_env`, that would mean we invoke the query to evaluate the static
+        // with different parameter environments, thus causing the static to be evaluated
+        // multiple times.
```

Comment:
Uh... is this really the right place to do such an optimization?

Reply:
Oh and now we also have this code duplicated between `eval_context` and `operand`. :/

Maybe at least have both of these copies right next to each other here in `eval_context`?

Reply:
> Uh... is this really the right place to do such an optimization?

No, but https://github.com/rust-lang/rust/pull/66877 already addresses this so I'm not gonna change it here

Reply:
And turns out the existing `const_eval_raw` in `eval_context.rs` also does this.

So, I think the cleanest approach is to also have a `const_eval` in `eval_context.rs`, mirroring the `const_eval_raw` there.

Though I still don't understand what exactly is prompting the change from `const_eval_raw` to `const_eval` here. Is one of the test-cases affected by this?

Reply:
> Though I still don't understand what exactly is prompting the change from const_eval_raw to const_eval here. Is one of the test-cases affected by this?

I don't remember. @spastorino do you remember? Something about a failing test, but I don't know if we were missing validation or if it was something else

Reply:
@oli-obk what about this proposal?

> So, I think the cleanest approach is to also have a const_eval in eval_context.rs, mirroring the const_eval_raw there.

Reply:
I added this. That did have a nice effect indeed